Perf improvement of Conv2d and Conv3d

lly-zero-one opened the pull request:
Summary: 1) avoid the use of item 2) bypass the im2col for 1x1 conv

Test Plan: unit test and perf benchmark to show improvement(WIP)

Differential Revision: D22149067

allwu commented on aten/src/ATen/native/ConvolutionMM3d.cpp:
```diff
@@ -189,6 +189,19 @@ static void slow_conv3d_update_output_frame(
     int64_t output_depth,
     int64_t output_height,
     int64_t output_width) {
+  if ((kernel_depth == 1) && (pad_depth == 0) && (stride_depth == 1) && (stride_height == 1) && (stride_width == 1) && (groups == 1)) {
```
Don't we need to check width/height for kernel/pad also?

allwu commented on aten/src/ATen/native/ConvolutionMM3d.cpp:
```diff
@@ -224,9 +237,7 @@ static void slow_conv3d_update_output_frame(
-      for (int64_t i = 0; i < n_output_plane; ++i) {
-        output[i].fill_(bias[i].item());
-      }
+      output.copy_(bias.unsqueeze(-1).unsqueeze(-1).unsqueeze(-1));
```
This is nice! This avoids the select op, right?

lly-zero-one replied:
Yes, credit goes to @ngimel

lly-zero-one committed:
Perf improvement of Conv2d and Conv3d (#40324)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/40324

1) avoid the use of item 2) bypass the im2col for 1x1 conv

Test Plan:
unit test and perf benchmark to show improvement
```
num = 50

N = 1
C = 512
H = 4
W = 4

M = 512
kernel_h = 1
kernel_w = 1
stride_h = 1
stride_w = 1
padding_h = 0
padding_w = 0

X_np = np.random.randn(N, C, H, W).astype(np.float32)
W_np = np.random.randn(M, C, kernel_h, kernel_w).astype(np.float32)
X = torch.from_numpy(X_np)

conv2d_pt = torch.nn.Conv2d(
    C, M, (kernel_h, kernel_w), stride=(stride_h, stride_w),
    padding=(padding_h, padding_w), groups=1, bias=True)

class ConvNet(torch.nn.Module):
    def __init__(self):
        super(ConvNet, self).__init__()
        self.conv2d = conv2d_pt

    def forward(self, x):
        return self.conv2d(x)

model = ConvNet()

def pt_forward():
    # with torch.autograd.profiler.profile(record_shapes=True) as prof:
    model(X)
    # print(prof.key_averages().table(sort_by="self_cpu_time_total"))

torch._C._set_mkldnn_enabled(False)

t = Timer("pt_forward()", "from __main__ import pt_forward, X")
```
Before the optimization:
pt time = 5.841153813526034
After the optimization:
pt time = 4.513134760782123

Differential Revision: D22149067

fbshipit-source-id: 7532eb9ffc57c9bc6cc3c95964d8d4c698a83ce8

allwu reviewed (approved):
LGTM!
